src/vm/method.hpp

```diff
@@ -258,6 +258,17 @@ class MethodDesc
         return (m_bFlags2 & enum_flag2_HasStableEntryPoint) != 0;
     }
 
+    inline BOOL HasStableEntryPoint1()
+    {
+        LIMITED_METHOD_DAC_CONTRACT;
+
+#if defined(FEATURE_JIT_PITCHING)
+        if (IsPitchable() && HasTemporaryEntryPoint())
+           return FALSE;
+#endif
+        return HasStableEntryPoint();
+    }
+
```

Comment:
This still feels like it adds a lot of confusion about what it means for the code pointer to be stable so unfortunately I'm also not a big fan of this. Looking at the problem more I've only come up with one solution so far that seems decent: using a fixup precode. If after reading through all of this you object then I'm happy to brainstorm further with you, but there are a few reasons why I think its a good choice:
a) It will align with concepts already in the code making it much easier to understand, and more likely to 'just work' when someone else changes code here without explicitly retesting or reasoning about code pitching.
b) It reuses machinery we have used in other scenarios for a long time, so you get the benefit of that in the field testing. Its probably going to be more reliable that the scheme we've got here and I am not aware of any known issues.
c) It should be easy to convert from where you are now to the fixup prestub.

In terms of what that solution looks like, not sure if you have familiarized yourself with the concept already but if not a fixup precode is a little fragment of code that can be rewritten as needed. It can be used as the stable entrypoint of a method regardless whether that method has code jitted or not. Initially the fragment of code looks like a single instruction "call PrecodeFixupThunk". When that code is called it will run PrecodeFixupThunk, which then calls ThePreStub, PreStubWorker, etc. After the code is generated the fixup prestub will be backpatched so that it now reads "jmp <addr_of_jitted_code>". You can also call Reset (ie https://github.com/dotnet/coreclr/blob/master/src/vm/method.cpp#L2483) to revert back to the state where the method will call the prestub again.

In practical terms to use it I think you only need to do this:
a) Add IsPitchable() to the list of conditions that make IsNativeCodeStableStableAfterInit() return false. That will cause DoPrestub() to automatically allocate and use the precode for you: https://github.com/dotnet/coreclr/blob/master/src/vm/prestub.cpp#L1370
https://github.com/dotnet/coreclr/blob/master/src/vm/prestub.cpp#L1531

b) When using a precode, you will need to have to have allocated a NativeCodeSlot. This where we hold the pointer to the jitted code because the stable entrypoint pointer has now been used to point at the precode. Update NeedsNativeCodeSlot to return TRUE for pitchable or potentially pitchable methods:
https://github.com/dotnet/coreclr/blob/master/src/vm/methodtablebuilder.cpp#L7044

c) In your new PitchNativeCode function, call Precode::Reset() to point the precode back to the prestub. You would no longer clear the stable entrypoint or clear the HasStableEntrypoint bit. You can call GetNativeCode() to get the previous jitted code pointer, use the current FreeCodeMemory function to free it, and then use SetNativeCodeInterlocked(NULL) to clear the slot. Technically you probably still have narrow race conditions with some other code in prestub, but even if you don't want to do the careful analysis to find them all its probably a much smaller set of potential races than the current mechanism is exposed to.